### PR TITLE
run on green waterfall revisions only

### PR DESCRIPTION
**Pull request by yjbanov**
Fixes https://github.com/flutter/flutter/issues/3776

/cc @devoncarew

### CODE REVIEW

File: bin/build.dart [hunk 1]
```diff
@@ -39,13 +40,32 @@ Future<Null> main(List<String> args) async {
 }
 
 Future<Null> build() async {
+  String revision = await getLatestGreenRevision();
+  if (ranBefore(revision)) {
```
Comment (devoncarew) [inline bin/build.dart]: `ranBefore` ==> `hasAlreadyRun`?

File: lib/src/buildbot.dart [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+///           "name2",
+///           value2,
+///           ... things we don't care about ...
+///         ]
+///       ]
+///     }
+dynamic _getBuildProperty(Map<String, dynamic> buildJson, String propertyName) {
+  List<List<dynamic>> properties = buildJson['properties'];
+  for (List<dynamic> property in properties) {
+    if (property[0] == propertyName) {
```
Comment (devoncarew) [inline lib/src/buildbot.dart]: nit: unnecessary braces

**Comment by devoncarew** [comment]
Do you want to store the commit markers files in `Directory.systemTemp.path`? I believe the temp dir will be managed outside of our control. Perhaps in `$HOME/.perfbot/` or some such?

lgtm w/o changing the commit dir location, if you think system temp would work

File: bin/build.dart (hunk 1 above)
Comment (yjbanov) [inline bin/build.dart]: Done.

File: lib/src/buildbot.dart (hunk 2 above)
Comment (yjbanov) [inline lib/src/buildbot.dart]: Done

**Comment by yjbanov** [comment]
> Perhaps in $HOME/.perfbot/ or some such?

Done: `$HOME/.flutter_dashboard`

**Comment by devoncarew** [comment]
@yjbanov, I think we're missing a lot of commits from this change - commits that are green on both the linux and mac flutter repo bots. We may want to revisit the logic -

**Comment by yjbanov** [comment]
I can take a look. Which ones are missing? Also, how can you tell? :) The mac build was broken for [most of the day](https://build.chromium.org/p/client.flutter/builders/Mac) yesterday. We did submit [a few metrics](http://146.148.43.81:8080/Benchmark?benchmark=complex_layout_scroll_perf.average_frame_build_time_millis), which seems on the same order of magnitude as the number of green builds from the past two days.

**Comment by devoncarew** [comment]
The ones in the evening - where the build was green - didn't get built. The last run was yesterday @ 1:00.

**Comment by yjbanov** [comment]
It looks like the build is occasionally failing. One of the tests is timing out (takes >30s to complete). I think the issue is that the device I'm testing on is getting slower over time. I'll restart it and see if it helps. Not sure what the long-term solution is. Would be nice if we could reboot the device automatically every now and then.

**Comment by devoncarew** [comment]
This guy: `top -stats pid,command,cpu,th,pstate,time -l2 -u -n20` would find any runaway processes on the machine (like an analysis server that hadn't stopped). It might be good to print out before a benchmark run.

**Comment by Hixie** [comment]
If it's the device getting slower that wouldn't help, since we're not
running analyzer on the device. The device getting slower could be
something we're doing. Do we have multiple flutter processes running or
something?

**Pull request merged**